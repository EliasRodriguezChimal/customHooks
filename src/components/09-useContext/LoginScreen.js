import React, { useContext } from 'react';
import { UserContext } from './UserContext';

export const LoginScreen = () => {
    //TAREA

    //1. Obtener la referenia al Context
    //2. Extraer el setUser del Context
    const {setUser} = useContext(UserContext);

    return (
        <div>
            <h1>LoginScreen</h1>
            <hr />

            <button
                className="btn btn-primary"
                onClick={() => { setUser({
                    id: 123,
                    name: 'Elias'
                }) }}
            >
                Login
            </button>
        </div>
    )
}
