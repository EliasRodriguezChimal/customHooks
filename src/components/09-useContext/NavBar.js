import React from 'react';
import { NavLink } from 'react-router-dom';

export const NavBar = () => {
    return (
        <nav className="navbar navbar-expand-sm navbar-dark bg-dark">
            <a className="navbar-brand"> useContext </a>
            <div className="collapse navbar-collapse" id="navbarNavAltMarkup">
                <div className="navbar-nav">
                    <NavLink exact="true" className="nav-link" to="/"> Home </NavLink>
                    <NavLink exact="true" className="nav-link" to="/about"> About </NavLink>
                    <NavLink exact="true" className="nav-link" to="/login"> Login </NavLink>
                </div>
            </div>
        </nav>
    )
}
