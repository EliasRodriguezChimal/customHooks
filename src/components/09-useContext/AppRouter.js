//NOTA: Una aplicacion puede tener mas de un Router.

import React from 'react';
import {
    BrowserRouter as Router,
    Route,
    Routes
} from 'react-router-dom';
import { AboutScreen } from './AboutScreen'
import { HomeScreen } from './HomeScreen';
import { LoginScreen } from './LoginScreen';
import { NavBar } from './NavBar';

export const AppRouter = () => {
    return (
        <Router>
            <div>
                <NavBar />

                <div className="container">
                    {/* La version 6 de react-router-dom cambio 'Switch' por 'Routes' */}
                    <Routes>
                        {/* La variable 'path' determina el valor que debe tener la URL para renderizar en pantalla el
                    componente determinado por la variable 'element'. 'element' debe tener un valor igual a un componente,
                    como se puede observar abajo, incluyendo los simbolos '<  />' */}
                        <Route exact path="/" element={<HomeScreen />} />
                        <Route exact path="/about" element={<AboutScreen />} />
                        <Route exact path="/login" element={<LoginScreen />} />

                        {/* La propiedad 'exact' escrita arriba, le indica al codigo que la ruta debe ser exactamente igual a la
                    descrita en la variable 'path' para renderizar el componente descrito en 'element', ya que de no hacerse 
                    asi y ubicar, por ejemplo, en primer lugar el path="/" y abajo los demas, aunque nos encontremos en /login, 
                    se renderizara el componente HomeScreen ya que es el primero que se encontro que coincidiera (con su '/')
                    aunque no sea exactamente igual. */}

                        {/* La siguiente ruta (Route) es usada como el 'default' de un Switch, para ejecutarse cuando el valor del path
                    no haya coincidido con ningun otro valor (que no exista dicho path) */}
                        <Route path="*" element={<HomeScreen />} />
                    </Routes>
                </div>
            </div>
        </Router>
    )
}
