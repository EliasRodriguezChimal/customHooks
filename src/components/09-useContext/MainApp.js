import React, { useState } from 'react';
import { AppRouter } from './AppRouter';
import { UserContext } from './UserContext';

export const MainApp = () => {
    const [user, setUser] = useState({});

    //El UserContext nos servira para usar el valor pasado como parametro (en este caso 'user' y 'setUser') si y solo si
    //el componente que intenta usarlo se ubica dentro del UserContext.Provider
    return (
        <UserContext.Provider value={{user, setUser}}>
            <AppRouter />
        </UserContext.Provider>
    )
}
