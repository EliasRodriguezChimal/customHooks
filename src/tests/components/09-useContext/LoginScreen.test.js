import React from 'react';
import { mount } from 'enzyme';
import { LoginScreen } from '../../../components/09-useContext/LoginScreen';
import { UserContext } from '../../../components/09-useContext/UserContext';

describe('<LoginScreen /> test', () => {
    const setUser = jest.fn();
    const wrapper = mount(//the property passed should be ALWAYS 'value' = 'the value needed for the component'
    //and has to be passed with double '{' like below.
        <UserContext.Provider value={{setUser}}> 
            <LoginScreen />
        </UserContext.Provider>
    );

    test('should be displayed correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('should execute setUser with the expected argument', () => {
        wrapper.find('button').simulate('click');
        expect(setUser).toHaveBeenCalledWith({
            id: 123,
            name: 'Elias'
        });
    });
})