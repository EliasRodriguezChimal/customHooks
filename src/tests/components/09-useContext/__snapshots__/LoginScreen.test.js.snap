// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginScreen /> test should be displayed correctly 1`] = `
<div>
  <h1>
    LoginScreen
  </h1>
  <hr />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;
